recipes/cad1/task1/baseline/enhance.py

```diff
@@ -0,0 +1,309 @@
+    """Scale signals back to the original scale."""
+    return sources * ref.std() + ref.mean()
+
+
+def map_to_dict(sources: np.ndarray, sources_list: List[str]) -> Dict:
+    """Map sources to a dictionary."""
+    audios = dict(zip(sources_list, sources))
+
+    signal_stems = {}
+    for source in sources_list:
+        audio = audios[source]
+        signal_stems[f"l_{source}"] = audio[0]
+        signal_stems[f"r_{source}"] = audio[1]
+
+    return signal_stems
+
+
+def decompose_signal(
+    model: torch.nn.Module, signal: np.ndarray, fs: int, device: torch.device
+) -> Dict[str, np.ndarray]:
+    """Decompose signal into 8 stems."""
+    signal, ref = normalize_signal(signal)
+    sources = separate_sources(model, signal, fs, device=device)
+    # only one element in the batch
+    sources = sources[0]
+    sources = denormalize_signals(sources, ref)
+    signal_stems = map_to_dict(sources, model.sources)
+    return signal_stems
+
+
+def apply_baseline_ha(
+    enhancer,
+    compressor,
+    signal: np.ndarray,
+    listener_audiogram: np.ndarray,
+    cfs: np.ndarray,
+) -> np.ndarray:
+    """
+    Apply NAL-R prescription hearing aid to a signal.
+
+    Args:
+        enhancer: An object that enhances the signal.
+        compressor: An object that compresses the signal.
+        signal: An ndarray representing the audio signal.
+        listener_audiogram: An ndarray representing the listener's audiogram.
+        cfs: An ndarray of center frequencies.
+
+    Returns:
+        An ndarray representing the processed signal.
+    """
+    nalr_fir, _ = enhancer.build(listener_audiogram, cfs)
+    proc_signal = enhancer.apply(nalr_fir, signal)
+    proc_signal, _, _ = compressor.process(proc_signal)
+    return proc_signal
+
+
+def process_stems_for_listener(
+    stems: dict,
+    enhancer,
+    compressor,
+    audiogram_left: np.ndarray,
+    audiogram_right: np.ndarray,
+    cfs: np.ndarray,
+) -> dict:
+    """Process the stems from sources.
+
+    Args:
+        stems (dict) : Dictionary of stems
+        enhancer (NALR) : NAL-R prescription hearing aid
+        compressor (Compressor) : Compressor
+        audiogram_left (np.ndarray) : Left channel audiogram
+        audiogram_right (np.ndarray) : Right channel audiogram
+        cfs (np.ndarray) : Center frequencies
+
+    Returns:
+        processed_sources (dict) : Dictionary of processed stems
+    """
+
+    processed_stems = {}
+
+    for stem_str in stems:
+        stem_signal = stems[stem_str]
+
+        # Determine the audiogram to use
+        audiogram = audiogram_left if stem_str.startswith("l") else audiogram_right
+
+        # Apply NALR prescription to stem_signal
+        proc_signal = apply_baseline_ha(
+            enhancer, compressor, stem_signal, audiogram, cfs
+        )
+        processed_stems[stem_str] = proc_signal
+    return processed_stems
+
+
+@hydra.main(config_path="", config_name="config")
+def enhance(config: DictConfig) -> None:
```

Comment:
This is quite a large function, it could perhaps be a `class` (e.g. `Enhance`) and have a number of smaller methods written for it. This would facilitate writing unit-tests to ensure each step runs correctly and a final method `.enhance()` that runs all steps in the required order.